rig/bitfield.py

```diff
@@ -487,156 +517,374 @@ def __init__(self, length=None, start_at=None, tags=None,
+            pairs.
+
+            This iterator includes fields which may not yet be accessible (but
+            which could be with additional field values set) but not fields
+            which require a different set of field_values set. Iteration
+            proceeds in a depth-first fashion with the ordering of each level
+            of the hierarchy matching insertion order.
+            """
+            for identifier, field in iteritems(self.fields):
                 yield (identifier, field)
-            else:
-                blocked.add(identifier)
 
-    def _assign_enabled_fields_with_fixed_start(self):
-        """For internal use only. Assign a length to any enabled fields which
-        do not have one but do have a fixed position.
-        """
-        assigned_bits = 0
-        unassigned_fields = list()
-        for identifier, field in self._enabled_fields():
-            if field.length is not None and field.start_at is not None:
-                assigned_bits |= ((1 << field.length) - 1) << field.start_at
-            elif field.start_at is not None:
-                unassigned_fields.append((identifier, field))
+            for requirements, child in self._potential_children(field_values):
+                for identifier, field in child.potential_fields(field_values):
+                    yield (identifier, field)
 
-        for identifier, field in unassigned_fields:
-            assigned_bits = self._assign_enabled_field(
-                assigned_bits, identifier, field)
+    def _assign_fields(self, identifiers, field_values,
+                       assign_positions, assigned_bits=0):
+        """For internal use only.  Assign lengths & positions to a subset of all
+        potential fields with the supplied field_values.
 
-    def _assign_enabled_fields_without_fixed_start(self, assigned_bits,
-                                                   exclude_fields=set()):
-        """For internal use only. Assign a length and position to any enabled
-        fields which do not have a fixed position.
+        This method will check for any assigned bits of all potential fields
+        but will only assign those fields whose identifiers are provided.
 
         Parameters
         ----------
-        exclude_fields : {identifier, ...}
-            Exclude fields which will be assigned later, for example those
-            which will be assigned by bitfields higher up the hierarchy.
+        identifiers : iterable of identifiers
+            The identifiers of the fields to assign
+        field_values : {identifier: value, ...}
+            The values held by various fields (used to access the correct
+            identifiers)
+        assign_positions : bool
+            If False, will only assign fields whose length is unknown (leaving
+            those with undefined positions as they are). If True, will assign
+            positions and lengths as required.
```

Comment:
I find the first sentence of this docstring very unclear.

"If False, will only assign lengths to fields whose positions are already known.  Otherwise lengths and positions will be assigned to all fields as necessary." Perhaps?